File: preprocessing/expandfeatures.py
import os, sys
from itertools import combinations



class ExpandFeatures():
    NUM_CNF = 5

    def __init__(self, cnfs, file, file_ex):
        cnfs = [cnf for cnf in map(int, str(cnfs))]

        if len(cnfs) != self.NUM_CNF:
            raise Exception("Invalid cnfs") 
        
        self.cnfs = dict()
        self.cnfs["x_or_y"] = cnfs[0]
        self.cnfs["notx_or_y"] = cnfs[1]
        self.cnfs["x_or_noty"] = cnfs[2]
        self.cnfs["notx_or_noty"] = cnfs[3]
        self.cnfs["x_xor_y"] = cnfs[4]
        self.expand_features(file, file_ex)


    def rawcount(self, file):
        f = open(file, "rb")
        lines = 0
        buf_size = 1024 * 1024
        read_f = f.raw.read
        buf = read_f(buf_size)

        while buf:
            lines += buf.count(b"\n")
            buf = read_f(buf_size)

        return lines


    def expand_features(self, file, file_ex):
        fex = open(file_ex, "a")
        num_lines = self.rawcount(file)

        with open(file, "r") as f:
            for cnt, line in enumerate(f):
                line = [int(i) for i in line.split(" ")]

                if cnt == 0:
                    line_size = len(line)
                    features = [i for i in range(line_size)][1:]

                for f1, f2 in combinations(features, 2):
                    if self.cnfs["x_or_y"]:
                        x_or_y = line[f1] or line[f2]
                        line.append(int(x_or_y))
                    
                    if self.cnfs["notx_or_y"]:
                        notx_or_y = not line[f1] or line[f2]
                        line.append(int(notx_or_y))
                    
                    if self.cnfs["x_or_noty"]:
                        x_or_noty = line[f1] or not line[f2]
                        line.append(int(x_or_noty))
                    
                    if self.cnfs["notx_or_noty"]:
                        notx_or_noty = not line[f1] or not line[f2]
                        line.append(int(notx_or_noty))
                    
                    if self.cnfs["x_xor_y"]:
                        x_xor_y = line[f1] ^ line[f2]
                        line.append(int(x_xor_y))

                fex.write(" ".join(map(str, line))+"\n")
                file_size = "{:.2f}".format(os.stat(file_ex).st_size/1024 ** 2)
                print(f"\rProgress: {int(((cnt+1)/num_lines)%100 * 100)}% | File size: {file_size} MiB", end="", flush=True)

        fex.close()



if __name__ == "__main__":
    cnfs = sys.argv[1]
    file = sys.argv[2]
    file_ex = sys.argv[3]
    ExpandFeatures(cnfs, file, file_ex)
    

File: preprocessing/expandfeatures.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <filesystem>
#include <algorithm>
#include <chrono>
#include <string>
#include <vector>
#include <cmath>
#include <map>

using namespace std;

/* returns C(n,k), wrong result if bigger than int32 */
unsigned binom(unsigned n, unsigned k) {
    if (k > n) return 0;
    if (k * 2 > n) k = n-k;
    if (k == 0) return 1;

    int result = n;

    for(int i = 2; i <= k; ++i) {
        result *= (n-i+1);
        result /= i;
    }

    return result;
}


int* combinations(int n, int k) {
    string bitmask(k, 1);
    bitmask.resize(n, 0);
    int size = binom(n, k) * k;
    int* combinations = new int[size];
    int cnt = 0;

    do {
        for (int i=0; i<n; ++i) {
            if (bitmask[i]) {
                combinations[cnt] = i+1;
                cnt++;
            }
        }
    }
    while (prev_permutation(bitmask.begin(), bitmask.end()));

    return combinations;
}


class ExpandFeatures {
    const int CNFS = 5;
    const int COMB = 2;

    public:
    map<string, int> cnfsMap;

    ExpandFeatures(string cnfs, string file, string fileEx) {
        if (cnfs.size() == CNFS) {
            cnfsMap = {
            {"x_or_y" , int(cnfs[0]-'0')},
            {"notx_or_y" , int(cnfs[1]-'0')},
            {"x_or_noty" , int(cnfs[2]-'0')},
            {"notx_or_noty" , int(cnfs[3]-'0')},
            {"x_xor_y" , int(cnfs[4]-'0')},
            };
        }
        else {
            throw runtime_error("Invalid CNFs");
        }

        expandFeatures(file, fileEx);
    }


    int* parseLine(string lineStr, int size) {
        int* line = new int[size];
        string s;
        stringstream ss(lineStr);
        int cnt = 0;

        while (getline(ss, s, ' ')) {
            line[cnt] = stoi(s);
            cnt++;
        }

        return line;
    }


    int countLines(string file) {
        ifstream f(file);
        int numLines = count_if(istreambuf_iterator<char>{f}, {}, [](char c) { return c == '\n'; });
        f.close();

        return numLines;
    }


    int countFeatures(string file) {
        ifstream f(file);
        string lineStr;
        getline(f, lineStr);

        string s;
        stringstream ss(lineStr);
        vector<int> firstLine;

        while (getline(ss, s, ' ')) {
            firstLine.push_back(stoi(s));
        }

        return firstLine.size() - 1;
    }


    void expandFeatures(string file, string fileEx) {
        int numCnfs = count_if(cnfsMap.begin(), cnfsMap.end(), [&](const pair<string, int>& pair) { return pair.second == 1; });
        int numLines = countLines(file);
        int numFeatures = countFeatures(file);
        int* comb = combinations(numFeatures, COMB);
        int size = binom(numFeatures, COMB);
        ifstream dataset(file);
        ofstream datasetEx(fileEx);
        string lineStr;
        int cnt = 0;
        
        while (getline(dataset, lineStr)) {
            int lineSize = 1 + numFeatures + numCnfs * size;
            int* line = parseLine(lineStr, lineSize);

            for (int i=0; i<size; ++i) {
                int f1 = comb[i*COMB];
                int f2 = comb[i*COMB+1];
                int index = 1 + numFeatures + numCnfs * i;
                int k = 0;

                if (cnfsMap["x_or_y"]) {
                    line[index + k] = line[f1] || line[f2];
                    k++;
                }

                if (cnfsMap["notx_or_y"]) {
                    line[index + k] = !line[f1] || line[f2];
                    k++;
                }

                if (cnfsMap["x_or_noty"]) {
                    line[index + k] = line[f1] || !line[f2];
                    k++;
                }

                if (cnfsMap["notx_or_noty"]) {
                    line[index + k] = !line[f1] || !line[f2];
                    k++;
                }

                if (cnfsMap["x_xor_y"]) {
                    line[index + k] = line[f1] ^ line[f2];
                }
            }
            
            for (int i=0; i<lineSize; ++i) {
                datasetEx << line[i] << ' ';
            }
            
            delete[] line;
            datasetEx << "\n";
            int progress = fmod(float(cnt+1)/float(numLines), 100) * 100;
            float file_size = float(filesystem::file_size(fileEx)) / (1024 * 1024);
            cout << "\r" << "Progress: " << progress << "%" << " | " << "File size: " << file_size << " MiB";
            cout.flush();
            cnt++;
        }
        
        delete[] comb;
        dataset.close();
        datasetEx.close();
    }
};


int main(int argc, char* argv[]) {
    string cnfs;
    string file;
    string fileEx;

    if (argc == 4) {
        cnfs = argv[1];
        file = argv[2];
        fileEx = argv[3];

        if (!filesystem::exists(file)) {
            throw runtime_error("File not found");
        }
    }
    else {
        throw runtime_error("Invalid arguments");
    }
    
    auto start = chrono::steady_clock::now();
    ExpandFeatures(cnfs, file, fileEx);
    auto end = chrono::steady_clock::now();
    
    auto execTime = end - start;

    cout << endl << chrono::duration <double, milli> (execTime).count() << " ms" << endl;

    return 0;
}
